src/main/java/rx/observables/AsyncOnSubscribe.java

```diff
@@ -0,0 +1,510 @@
+     *            {@link #next(Object, long, Observer) next(S, long, Observer)})
+     * @param onUnsubscribe
+     *            clean up behavior (see {@link #onUnsubscribe(Object) onUnsubscribe(S)})
+     * @return an OnSubscribe that emits data downstream in a protocol compatible with
+     *         back-pressure.
+     */
+    @Experimental
+    public static <T> OnSubscribe<T> createStateless(final Action2<Long, ? super Observer<Observable<? extends T>>> next, 
+            final Action0 onUnsubscribe) {
+        Func3<Void, Long, Observer<Observable<? extends T>>, Void> nextFunc = 
+                new Func3<Void, Long, Observer<Observable<? extends T>>, Void>() {
+                    @Override
+                    public Void call(Void state, Long requested, Observer<Observable<? extends T>> subscriber) {
+                        next.call(requested, subscriber);
+                        return null;
+                    }};
+        Action1<? super Void> wrappedOnUnsubscribe = new Action1<Void>() {
+            @Override
+            public void call(Void t) {
+                onUnsubscribe.call();
+            }};
+        return new AsyncOnSubscribeImpl<Void, T>(nextFunc, wrappedOnUnsubscribe);
+    }
+
+    /**
+     * An implementation of AsyncOnSubscribe that delegates
+     * {@link AsyncOnSubscribe#next(Object, long, Observer)},
+     * {@link AsyncOnSubscribe#generateState()}, and {@link AsyncOnSubscribe#onUnsubscribe(Object)}
+     * to provided functions/closures.
+     *
+     * @param <S>
+     *            the type of the user-defined state
+     * @param <T>
+     *            the type of compatible Subscribers
+     */
+    private static final class AsyncOnSubscribeImpl<S, T> extends AsyncOnSubscribe<S, T> {
+        private final Func0<? extends S> generator;
+        private final Func3<? super S, Long, ? super Observer<Observable<? extends T>>, ? extends S> next;
+        private final Action1<? super S> onUnsubscribe;
+
+        private AsyncOnSubscribeImpl(Func0<? extends S> generator, Func3<? super S, Long, ? super Observer<Observable<? extends T>>, ? extends S> next, Action1<? super S> onUnsubscribe) {
+            this.generator = generator;
+            this.next = next;
+            this.onUnsubscribe = onUnsubscribe;
+        }
+
+        public AsyncOnSubscribeImpl(Func0<? extends S> generator, Func3<? super S, Long, ? super Observer<Observable<? extends T>>, ? extends S> next) {
+            this(generator, next, null);
+        }
+
+        public AsyncOnSubscribeImpl(Func3<S, Long, Observer<Observable<? extends T>>, S> next, Action1<? super S> onUnsubscribe) {
+            this(null, next, onUnsubscribe);
+        }
+
+        public AsyncOnSubscribeImpl(Func3<S, Long, Observer<Observable<? extends T>>, S> nextFunc) {
+            this(null, nextFunc, null);
+        }
+
+        @Override
+        protected S generateState() {
+            return generator == null ? null : generator.call();
```

Comment:
It would be great, if this is to be migrated to 2.x, that user-supplied functions are called in try-catch. Perhaps not here but in call().